### PR TITLE
Customise view, edit and source links through annotations

### LINKED ISSUE #4455: Allow custom edit/view links in CatalogTable rows
## Feature Suggestion

The catalog table includes edit and view links. Make these customisable on a per location basis.

## Possible Implementation

This could be added to the various schemas for locations, in app-config.yaml for example:

```yaml
locations:
  - type: url
    target: https://github.com/example/example/blob/master/catalog-info.yaml
    editUrl: https://github.com/example/example/edit/master/catalog/index.jsonnet
```

The explicit URLs would then override the link builder functions in the catalog plugin.

```ts
export const createEditLink = (location: LocationSpec): string | undefined => {
  if (location.editUrl) return location.editUrl;
  // ... existing implementation goes here ... 
}
```

## Context

The current implementation works reliably if:

1. You are using one of the supported Git providers to store your catalog manifests
2. The YAML file itself is the source of truth, ie, it is not generated from other source files

If the first condition is not true, then edit/view links may or may not take the user to the right place. For example, we have a custom catalog processor that reads from a GCS bucket, and in that case the links take you to a browser-unfriendly `gs://` url. 

If the second condition is not true, then you may inadvertently send users to edit generated files, which is not ideal. 

Allowing customisation would prevent these problems. Doing it per location would seem to be the best idea since these links are tied to the particular implementation of the location's target.

### PR DESCRIPTION
**Pull request by tragiclifestories**
Fixes #4455.

## Hey, I just made a Pull Request!

Allow custom links on catalog and entity pages to entity metadata and source. So if a component has the following annotations:

```yaml
metadata:
  annotations:
    backstage.io/browser-view-url: https://some.website/catalog-info.yaml
    backstage.io/browser-edit-url: https://github.com/my-org/catalog/edit/master/my-service.jsonnet
    backstage.io/browser-source-url: https://github.com/my-org/my-service
```
These links in the Actions column on the CatalogPage:

2021-02-18-143548_3840x1080_scrot

... will point to `https://some.website/catalog-info.yaml` and `https://github.com/my-org/catalog/edit/master/my-service.jsonnet`, and on the entity's own page:
 
2021-02-18-143626_3840x1080_scrot

... the View source link on the left, and the edit link on the right, will point to `https://github.com/my-org/my-service` and `https://github.com/my-org/catalog/edit/master/my-service.jsonnet`.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by tragiclifestories** [comment]
> message: 'Did you really mean \'Jsonnet\'?'

Yes I did :sweat_smile: - not sure what the remediation is here ...

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4575-34xd2ktpkpef5sjagimdsommbwb7y2br.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602e7caff285af104b92040d

File: plugins/catalog/src/components/AboutCard/AboutCard.tsx
Comment (adamdmharvey) [inline plugins/catalog/src/components/AboutCard/AboutCard.tsx:79]: Feel like there's some tweaks to the logic needed here.  For example, even if there's a source URL annotation, if there is no location, this will return blank?  I don't recall, will location ALWAYS be valid?  if so, maybe the if source work should be moved in there, OR maybe this should be a broader with psuedocode like:

```
if sourceUrl annotation has been specified
  use it for source link
else
  try to use the location target for source link
end
return source link
```

Not sure if that makes sense 😁

Comment (tragiclifestories) [reply plugins/catalog/src/components/AboutCard/AboutCard.tsx:79]: The complicating factor is the edithref which depends on the location, but actually I think that can be dealt with.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4575-34xd2ktpkpef5sjagimdsommbwb7y2br.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602e7caff285af104b92040d

File: docs/features/software-catalog/well-known-annotations.md
Comment (Rugvip) [inline docs/features/software-catalog/well-known-annotations.md:81]: @backstage/maintainers, thoughts on there? I'm thinking we drop `browser-`, e.g. `backstage.io/view-url`?

Full example:

```yaml
backstage.io/view-url: https://some.website/catalog-info.yaml
backstage.io/edit-url: https://github.com/my-org/catalog/edit/master/my-service.jsonnet
backstage.io/source-url: https://github.com/my-org/my-service
```

File: docs/features/software-catalog/well-known-annotations.md
Comment (Rugvip) [inline docs/features/software-catalog/well-known-annotations.md:89]: Too outlandish for Vale, gotta add it to `.github/styles/vocab.txt` :grin:

Or tbh you could also make this `.jsonnet` since code blocks aren't checked

**Review by Rugvip — COMMENTED** [review]
Thank you! 😁 

I think the idea and implementation is good, it's nice to have these optional, and typically they'd be generated by a catalog processor too.

Just gotta have a bit of bikeshedding over the naming :grin:

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4575-34xd2ktpkpef5sjagimdsommbwb7y2br.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602e7caff285af104b92040d

File: docs/features/software-catalog/well-known-annotations.md
Comment (tragiclifestories) [reply docs/features/software-catalog/well-known-annotations.md:89]: Oh that's where it is. Cheers!

File: docs/features/software-catalog/well-known-annotations.md
Comment (tragiclifestories) [reply docs/features/software-catalog/well-known-annotations.md:81]: I'd prefer the shorter version FWIW. I blindly copied the browser prefix from the issue but on reflection it only really makes sense for `edit` since that link literally edits a file in your browser.

**Comment by adamdmharvey** [comment]
BTW, one idea I've seen in a few places is where ever you're statically defining the annotation in code, would be to define that annotation in a constants file, something like this:

https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/ingestion/processors/ldap/constants.ts

That way changing it is relatively simple, and reduces the places where a mistake could be made.

File: docs/features/software-catalog/well-known-annotations.md
Comment (freben) [reply docs/features/software-catalog/well-known-annotations.md:81]: Yeah I think you are right. In particular the source one should be usable by other plugins as a proper pointer to the location of the source that the entity describes. I think it should even be an actual location, and named `source-location` symmetrically to the `managed-by-location`.

```yaml
backstage.io/source-location: url:https://github.com/my-org/my-service
```

As a side note, I think it's time we standardized on a location string parser that's smart enough to interpret `http(s):` as `url:http(s):`... :)

**Comment by Fox32** [comment]
We actual had the need to remove these buttons the other day. I wonder if we want to extend it, so that an empty annotation means that the button is hidden? Alternative would be to always add the annotations via a processor instead of falling back to the old logic and skip the button if no annotation is present. But that would be something for another PR.

**Comment by freben** [comment]
That's an interesting thought @Fox32 - I hadn't thought of it that way, that perhaps this could be entirely backend driven. That way, the frontend won't have to worry at all about how these URLs are formed.

File: docs/features/software-catalog/well-known-annotations.md
Comment (freben) [reply docs/features/software-catalog/well-known-annotations.md:81]: If we go with my suggestion on the latter (@Rugvip?) then maybe this section of the docs could be split apart into the view-url and edit-url in one heading, and the source-location in another.

Comment (tragiclifestories) [reply docs/features/software-catalog/well-known-annotations.md]: @freben @Rugvip I made that change to a proper location for source files in https://github.com/backstage/backstage/pull/4575/commits/e0fc8db66a79d5f236ef251b2effcc2b690b5f8c, can be reverted if we change our minds.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4575-34xd2ktpkpef5sjagimdsommbwb7y2br.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602e7caff285af104b92040d

**Comment by Fox32** [comment]
Don't want to block this PR, but I could offer implementing the existing frontend code in the backend and make the annotations "optional" (hide/disable the features in the UI) in a follow-up PR as we face a very similar issue right now.

**Comment by tragiclifestories** [comment]
> implementing the existing frontend code in the backend

Just so I'm clear, you'd mean changing the built-in location readers and/or processors to add in annotations for these things and remove all the link-builder gymnastics from the front-end plugin?

**Comment by Fox32** [comment]
Exactly, maybe in a new processor so that it's easy to override the behavior

**Comment by freben** [comment]
Hmm not really right? I look at this analyze endpoint in the backend to be a very standalone thing.

**Comment by Fox32** [comment]
You mean something like the location analyzer that is called from the frontend and not use annotations at all?

**Comment by freben** [comment]
Whoa, strike that, that comment was not meant for this PR at all :) Sorry for the confusion.

**Comment by Fox32** [comment]
Haha, nice

**Comment by adamdmharvey** [comment]
Note, regarding hiding buttons unless annotation exists if that's still in the "cards" (wacka wacka), if I put on my "Stefan Alund Hat" 🎩 I might roughly say, we may want to try to use the UI to drive _discovery_ without hiding features. Most users won't know they could add a link that might appear at the top of an About card, without reading through the the documentation or seeing another entity and then _figuring out_ how to customize an entity meta data file, versus seeing a link is disabled or missing, and discovering they may find value in it.  Just a thought!

**Comment by Fox32** [comment]
Yeah disabling it would be an alternative. Maybe a bit about the use case: there are a lot of adopters that fill the catalog from alternative sources or have a custom backend. In these cases the links are only partial useful or might be used to link to an external tool.

**Comment by freben** [comment]
@stefanalund is indeed well known for wearing golden-banded top hats

File: plugins/catalog/src/components/actions.ts
Comment (freben) [inline plugins/catalog/src/components/actions.ts:98]: So i'm thinking ... this logic (and same above),

Just move the `findLocationForEntityMeta` call in here, and remove the location argument. And since these are so browser specific, maybe we should only return something when that returns a `url` type and the target starts with `http:` or `https:`? Otherwise we can just return undefined.

File: plugins/catalog/src/components/CatalogTable/CatalogTable.tsx
Comment (freben) [inline plugins/catalog/src/components/CatalogTable/CatalogTable.tsx:170]: We mentioned disabling these when there was no url to go to, right? I am not sure how to best make it visually disabled, but atleast the tooltip should tell you why you can't click it. Same above

**Review by freben — COMMENTED** [review]
Nice. Two smaller comments

**Comment by freben** [comment]
Oh one more thing. I think we said to remove the "browser" prefix on the other two as well.

```yaml
backstage.io/view-url: ...
backstage.io/edit-url: ...
```

File: plugins/catalog/src/components/CatalogTable/CatalogTable.tsx
Comment (tragiclifestories) [reply plugins/catalog/src/components/CatalogTable/CatalogTable.tsx:170]: I think that as currently implemented, there will always be a URL, since if all else fails it will just fall back to `location.target`. `location` is hypothetically nullable, but I can't see how you'd get there without manually editing the database ... 

Cleanly fixing that I think would mean doing what @Fox32 suggests and moving the logic to the backend/processor step.

File: plugins/catalog/src/components/actions.ts
Comment (tragiclifestories) [reply plugins/catalog/src/components/actions.ts]: The only caveat would be that you're going to do the `findLocationForEntityMeta` twice on the AboutCard I think. But it shouldn't make any real difference.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4575-34xd2ktpkpef5sjagimdsommbwb7y2br.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602e7caff285af104b92040d

**Comment by tragiclifestories** [comment]
@freben Think that's everything.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4575-34xd2ktpkpef5sjagimdsommbwb7y2br.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602e7caff285af104b92040d

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4575-34xd2ktpkpef5sjagimdsommbwb7y2br.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602e7caff285af104b92040d

**Comment by tragiclifestories** [comment]
FOSSA test seems to be failing on upload ...

File: plugins/catalog/src/components/actions.ts
Comment (freben) [reply plugins/catalog/src/components/actions.ts]: No problem. And could be refactored to calculate both in one go if desired :)

File: plugins/catalog/src/components/CatalogTable/CatalogTable.tsx
Comment (freben) [reply plugins/catalog/src/components/CatalogTable/CatalogTable.tsx:167]: But location targets are not always even URLs at all. They could even be the empty string. We can make no assumptions about them whatsoever actually. So we could even go as far as running it through the URL() constructor and return undefined if it throws, or if the protocol was neither http nor https.

**Comment by freben** [comment]
Yeah don't worry about fossa, it's OK :)

**Review by freben — APPROVED** [review]
Let's get this in for now though, we can iterate on the details later. Thanks!

**Pull request merged**
